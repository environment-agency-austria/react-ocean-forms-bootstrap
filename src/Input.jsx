/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { Input as StrapInput } from 'reactstrap';
import { fieldMetaShape, fieldShape } from 'react-ocean-forms';

import FieldLine from './FieldLine';

/**
 * Component for displaying bootstrap
 * form groups with an html input and
 * oForm support
 */
function Input(props) {
  const {
    field,
    type,
    meta,
  } = props;

  const invalid = meta.valid === true ? undefined : true;

  return (
    <FieldLine {...props}>
      <StrapInput type={type} {...field} invalid={invalid} />
    </FieldLine>
  );
}

Input.defaultProps = {
  info: undefined,
  type: 'text',
};

Input.propTypes = {
  label: PropTypes.string.isRequired,
  info: PropTypes.string,
  type: PropTypes.string,
  meta: fieldMetaShape.isRequired,
  field: fieldShape.isRequired,
};

export default Input;
