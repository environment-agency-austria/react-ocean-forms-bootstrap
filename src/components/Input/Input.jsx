/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { Input as StrapInput } from 'reactstrap';
import { fieldMetaShape, fieldShape } from 'react-ocean-forms-legacy';

import { FieldLine } from '../FieldLine';

/**
 * Component for displaying bootstrap
 * form groups with an html input and
 * oForm support
 */
function Input(props) {
  const {
    field,
    type,
    meta,
  } = props;

  const invalid = meta.valid === true ? undefined : true;

  return (
    <FieldLine {...props}>
      <StrapInput type={type} {...field} invalid={invalid} plaintext={meta.plaintext}>
        {meta.plaintext ? field.value : undefined}
      </StrapInput>
    </FieldLine>
  );
}

Input.displayName = 'Input';

Input.defaultProps = {
  info: undefined,
  type: 'text',
};

Input.propTypes = {
  label: PropTypes.string.isRequired,
  info: PropTypes.string,
  type: PropTypes.string,
  meta: fieldMetaShape.isRequired,
  field: fieldShape.isRequired,
};

export default Input;
