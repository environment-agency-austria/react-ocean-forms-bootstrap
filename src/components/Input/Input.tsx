/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';

import { useField } from 'react-ocean-forms';
import { Input as StrapInput } from 'reactstrap';

import { FieldLine } from '../FieldLine';
import { IInputProps } from './Input.types';

/**
 * Component for displaying bootstrap
 * form groups with an html input and
 * react-ocean-forms support
 */
export const Input = <TSubmitValue extends unknown>(props: IInputProps<TSubmitValue>): JSX.Element => {
  const {
    type = 'text',
    ...rest
  } = props;

  const { fieldProps, metaProps } = useField(rest);

  const invalid = metaProps.valid === true ? undefined : true;
  const defaultizedValue = fieldProps.value ?? '';

  return (
    <FieldLine {...rest} fieldProps={fieldProps} metaProps={metaProps}>
      <StrapInput type={type} {...fieldProps} value={defaultizedValue} invalid={invalid} plaintext={metaProps.plaintext} />
    </FieldLine>
  );
};
