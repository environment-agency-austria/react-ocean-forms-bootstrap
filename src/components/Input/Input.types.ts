import { IUseFieldProps } from 'react-ocean-forms';
import { InputType } from 'reactstrap/lib/Input';

import { IBaseFieldLineProps } from '../FieldLine';

type DefaultHtmlInputValueType = React.InputHTMLAttributes<HTMLInputElement>['value'];

/**
 * Props for the Input component
 */
export interface IInputProps<TSubmitValue> extends IUseFieldProps<DefaultHtmlInputValueType, TSubmitValue>, IBaseFieldLineProps {
  /**
   * Html input type
   */
  type?: InputType;
}
