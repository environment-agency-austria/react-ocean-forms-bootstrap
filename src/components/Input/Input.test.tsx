import React from 'react';

import { render, fireEvent, waitForElement } from '@testing-library/react';
import { Form, validators } from 'react-ocean-forms';

import { Input } from './Input';

describe('<Input />', () => {
  it('should render correctly', () => {
    const { asFragment, getByRole, getByLabelText } = render(
      <Form>
        <Input name="mock" label="mock" />
      </Form>
    );

    expect(getByRole('textbox')).toBeVisible();
    expect(getByLabelText('mock')).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should let the user override the input type', () => {
    const inputType = 'number';
    const { getByLabelText } = render(
      <Form>
        <Input name="mock" label="mock" type={inputType} />
      </Form>
    );

    expect(getByLabelText('mock')).toBeVisible();
    expect(getByLabelText('mock')).toHaveAttribute('type', inputType);
  });

  it('should call field.onChange when the input changes', async () => {
    const changedValue = 'changed value';
    const handleChange = jest.fn();
    const { getByLabelText } = render(
      <Form>
        <Input name="mock" label="mock" onChange={handleChange} />
      </Form>
    );

    fireEvent.change(getByLabelText('mock'), { target: { value: changedValue }});
    expect(handleChange).toHaveBeenCalledWith(changedValue);

    const input = await waitForElement(() => getByLabelText('mock'));
    expect(input).toHaveValue(changedValue);
  });

  it('should call field.onBlur when there is an input blur', () => {
    const handleBlur = jest.fn();
    const { getByLabelText } = render(
      <Form>
        <Input name="mock" label="mock" onBlur={handleBlur} />
      </Form>
    );

    fireEvent.blur(getByLabelText('mock'));
    expect(handleBlur).toHaveBeenCalled();
  });

  it('should react on meta.plaintext correctly', () => {
    const { asFragment } = render(
      <Form plaintext>
        <Input name="mock" label="mock" />
      </Form>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should react on meta.valid correctly', () => {
    const { getByRole, getByLabelText, getByText } = render(
      <Form plaintext>
        <Input name="mock" label="mock" validators={[validators.required]} />
      </Form>
    );

    fireEvent.change(getByLabelText('mock'), { target: { value: 'temp value' }});
    fireEvent.change(getByLabelText('mock'), { target: { value: '' }});

    expect(getByRole('textbox')).toHaveClass('is-invalid');
    expect(getByText('This field is mandatory.')).toBeVisible();
  });
});
