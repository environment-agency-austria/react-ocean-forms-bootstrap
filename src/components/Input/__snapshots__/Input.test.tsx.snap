// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should react on meta.plaintext correctly 1`] = `
<FieldLine
  field={
    Object {
      "disabled": false,
      "id": "mock-field",
      "name": "field0",
      "onBlur": [MockFunction],
      "onChange": [MockFunction],
      "value": "",
    }
  }
  inputSize="9"
  label="field0"
  labelClass="text-right"
  labelSize="3"
  meta={
    Object {
      "error": null,
      "isRequired": false,
      "isValidating": false,
      "plaintext": true,
      "stringFormatter": [MockFunction],
      "touched": false,
      "valid": true,
    }
  }
  type="text"
>
  <Input
    disabled={false}
    id="mock-field"
    name="field0"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    plaintext={true}
    type="text"
    value=""
  />
</FieldLine>
`;

exports[`<Input /> should render correctly 1`] = `
<FieldLine
  field={
    Object {
      "disabled": false,
      "id": "mock-field",
      "name": "field0",
      "onBlur": [MockFunction],
      "onChange": [MockFunction],
      "value": "",
    }
  }
  inputSize="9"
  label="field0"
  labelClass="text-right"
  labelSize="3"
  meta={
    Object {
      "error": null,
      "isRequired": false,
      "isValidating": false,
      "plaintext": false,
      "stringFormatter": [MockFunction],
      "touched": false,
      "valid": true,
    }
  }
  type="text"
>
  <Input
    disabled={false}
    id="mock-field"
    name="field0"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    plaintext={false}
    type="text"
    value=""
  />
</FieldLine>
`;
